src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.commonjava.indy.service.archive.config.PreSeedConfig;
+import org.commonjava.indy.service.archive.model.dto.HistoricalContentDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+@ApplicationScoped
+public class ArchiveController {
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    PreSeedConfig preSeedConfig;
+
+    @Inject
+    ObjectMapper objectMapper;
+
+    private ExecutorService executorService;
+
+    private CloseableHttpClient client;
+
+    private String contentDir;
+
+    private String archiveDir;
+
+    private final String CONTENT_DIR = "/content";
+
+    private final String ARCHIVE_DIR = "/archive";
+
+    @PostConstruct
+    public void init()
+    {
+        int threads = 4 * Runtime.getRuntime().availableProcessors();
+        executorService = Executors.newFixedThreadPool( threads, ( final Runnable r ) -> {
+            final Thread t = new Thread( r );
+            t.setDaemon( true );
+            return t;
+        } );
+
+        final PoolingHttpClientConnectionManager ccm = new PoolingHttpClientConnectionManager();
+        ccm.setMaxTotal( 500 );
+
+        RequestConfig rc = RequestConfig.custom().build();
+        final HttpClientBuilder builder = HttpClients.custom().setConnectionManager( ccm ).setDefaultRequestConfig( rc );
+        client = builder.build();
+
+        String storeDir = preSeedConfig.storageDir.orElse( "data" );
+        contentDir = String.format( "%s%s", storeDir, CONTENT_DIR );
+        archiveDir = String.format( "%s%s", storeDir, ARCHIVE_DIR );
+    }
+
+    @PreDestroy
+    public void destroy()
+    {
+        IOUtils.closeQuietly( client, null );
+    }
+
+    public void downloadArtifacts( final Map<String, String> downloadPaths, final HistoricalContentDTO content )
+    {
+        BasicCookieStore cookieStore = new BasicCookieStore();
+        ExecutorCompletionService executor = new ExecutorCompletionService<>( executorService );
```

Comment:
Small culprit. Missed the Generic declaration for ExecutorCompletionService. (Should be ExecutorCompletionService<Boolean>?)

Reply:
Ok, this will be related to the executor result take, and asyn completion, I'll update in the amending commit together with other tiny places.